src/misc/unicode.cpp

```diff
@@ -0,0 +1,1159 @@
+static std::map<uint16_t, code_page_mapping_t> mappings_aliases = {};
+// Reverse mappings, codepage -> UTF-8
+static std::map<uint16_t, code_page_mapping_reverse_t> mappings_reverse = {};
+
+// ***************************************************************************
+// Grapheme type implementation
+// ***************************************************************************
+
+static bool is_combining_mark(const uint32_t code_point)
+{
+    static const std::vector<std::pair<uint16_t, uint16_t>> ranges = {
+        { 0x0300, 0x036f }, // Combining Diacritical Marks
+        { 0x1ab0, 0x1aff }, // Combining Diacritical Marks Extended
+        { 0x1dc0, 0x1dff }, // Combining Diacritical Marks Supplement
+        { 0x20d0, 0x20ff }, // Combining Diacritical Marks for Symbols
+        { 0xfe20, 0xfe2f }, // Combining Half Marks
+    };
+
+    for (const auto &range : ranges)
+        if (code_point >= range.first && code_point <= range.second)
+            return true;
+
+    return false;
+}
+
+Grapheme::Grapheme(const uint16_t code_point) :
+    code_point(code_point),
+    empty(false)
+{
+    // It is not valid to have a combining mark
+    // as a main code point of the grapheme
+
+    if (is_combining_mark(code_point))
+        Invalidate();
+}
+
+bool Grapheme::IsEmpty() const
+{
+    return empty;
+}
+
+bool Grapheme::IsValid() const
+{
+    return valid;
+}
+
+bool Grapheme::HasMark() const
+{
+    return mark_1st;
+}
+
+uint16_t Grapheme::GetCodePoint() const
+{
+    return code_point;
+}
+
+void Grapheme::Invalidate()
+{
+    empty = false;
+    valid = false;
+
+    code_point = unknown_character;
+    mark_1st   = 0;
+    mark_2nd   = 0;
+}
+
+void Grapheme::AddMark(const uint16_t code_point)
+{
+    if (!valid)
+        return; // can't add combining mark to invalid grapheme
+
+    if (!is_combining_mark(code_point) || empty) {
+        Invalidate(); // not a combining mark or empty grapheme
+        return;
+    }
+
+    if (mark_1st && mark_2nd)
+        return; // up to 2 combining marks supported, ignore the rest
+
+    if (mark_1st)
+        mark_2nd = code_point;
+    else
+        mark_1st = code_point;
+}
+
+void Grapheme::StripMarks()
+{
+    mark_1st = 0;
+    mark_2nd = 0;
+}
+
+bool Grapheme::operator<(const Grapheme &other) const
+{
+    // TODO: after migrating to C++20 use a templated lambda here
+
+    if (code_point < other.code_point)
+        return true;
+    else if (code_point > other.code_point)
+        return false;
+
+    if (mark_1st < other.mark_1st)
+        return true;
+    else if (mark_1st > other.mark_1st)
+        return false;
+
+    if (mark_2nd < other.mark_2nd)
+        return true;
+    else if (mark_2nd > other.mark_2nd)
+        return false;
+
+    if (empty < other.empty)
+        return true;
+    else if (empty > other.empty)
+        return false;
+
+    if (valid < other.valid)
+        return true;
+    else if (valid > other.valid)
+        return false;
```

Comment:
OK, changed. It doesn't matter if empty > non_empty or if empty < non_empty. What is important, is that it has to be a strict total order (otherwise the STL won't work correctly), which means:

1) A < A is always false
2) A < B and B < C implies that A < C
3) if A != B implies that either A < B or B < A

The logic here is similar to comparing two integers with the same amount of digits.